#!/usr/bin/env Rscript
#Written by Malachi Griffith

args = (commandArgs(TRUE))
working_dir = args[1];              #Expression matrix file of cufflinks expression values generated by buildCufflinksFpkmMatrix.pl
images_subdir = args[2];
fpkm_percent_cutoff = as.numeric(args[3]);


#working_dir = "/gscmnt/sata132/techd/mgriffit/hgs/all1/rnaseq/absolute/"
#images_subdir = "/gscmnt/sata132/techd/mgriffit/hgs/all1/rnaseq/absolute/images/"
#fpkm_percent_cutoff = 1

if (length(args) < 3){
  message_text1 = "Required arguments missing for outlierGenesAbsolute.R"
  stop(message_text1)
}

#Set the working directory
setwd(working_dir)

#Define the input file paths
genes_of_interest_file=paste(working_dir,"genes_of_interest.txt", sep="")
genes_fpkm_namesort_file=paste(working_dir,"genes.fpkm.namesort.tsv", sep="")
isoforms_fpkm_namesort_file=paste(working_dir,"isoforms.fpkm.namesort.tsv", sep="")
isoforms_merged_fpkm_namesort_file=paste(working_dir,"isoforms.merged.fpkm.namesort.tsv", sep="")

#Define the output file paths
#Sorted by expression level
genes_fpkm_expsort_file=paste(working_dir, "genes.fpkm.expsort.tsv", sep="")
isoforms_fpkm_expsort_file=paste(working_dir, "isoforms.fpkm.expsort.tsv", sep="")
isoforms_merged_fpkm_expsort_file=paste(working_dir, "isoforms.merged.fpkm.expsort.tsv", sep="")

#Sorted by expression level and filtered down to the top N% only
genes_fpkm_expsort_top_file=paste(working_dir, "genes.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".tsv", sep="")
isoforms_fpkm_expsort_top_file=paste(working_dir, "isoforms.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".tsv", sep="")
isoforms_merged_fpkm_expsort_top_file=paste(working_dir, "isoforms.merged.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".tsv", sep="")

#Sorted by expression level - genes of interest only
genes_fpkm_expsort_goi_file=paste(working_dir, "genes.fpkm.expsort.goi.tsv", sep="")
isoforms_fpkm_expsort_goi_file=paste(working_dir, "isoforms.fpkm.expsort.goi.tsv", sep="")
isoforms_merged_fpkm_expsort_goi_file=paste(working_dir, "isoforms.merged.fpkm.expsort.goi.tsv", sep="")

#Sorted by expression level and filtered down to the top N% only - genes of interest only
genes_fpkm_expsort_top_goi_file=paste(working_dir, "genes.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".goi.tsv", sep="")
isoforms_fpkm_expsort_top_goi_file=paste(working_dir, "isoforms.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".goi.tsv", sep="")
isoforms_merged_fpkm_expsort_top_goi_file=paste(working_dir, "isoforms.merged.fpkm.expsort.top", fpkm_percent_cutoff, "percent", ".goi.tsv", sep="")

#Load input files
goi = read.table(file=genes_of_interest_file, header=FALSE, as.is=1)
genes_fpkm = read.table(file=genes_fpkm_namesort_file, header=TRUE, as.is=c(1:4,10), na.strings=c("na","-","NA","N/A","n/a"))
isoforms_fpkm = read.table(file=isoforms_fpkm_namesort_file, header=TRUE, as.is=c(1:4,10), na.strings=c("na","-","NA","N/A","n/a"))
isoforms_merged_fpkm = read.table(file=isoforms_merged_fpkm_namesort_file, header=TRUE, as.is=c(1:4,10), na.strings=c("na","-","NA","N/A","n/a"))



writeFiles = function(data, outfile, outfile_goi, outfile_top, outfile_top_goi){
  #Write out the FPKM ordered files (with all data entries) - append a rank column as well
  o = order(data[,"FPKM"], na.last = TRUE, decreasing = TRUE)
  x = data[o,]
  x[,"rank"]=1:length(x[,1])
  x[,"percentile"]=round(((x[,"rank"]/length(x[,1]))*100), digits=3)
  write.table(x, file=outfile, quote = FALSE, sep="\t", row.names = FALSE)

  #Write out the subset of all genes that are in the GOI list
  goi_i=which(x[,"mapped_gene_name"] %in% goi[[1]])
  write.table(x[goi_i,], file=outfile_goi, quote = FALSE, sep="\t", row.names = FALSE)

  #Write out the top N% of these genes
  cutoff = length(x[,1]) * (fpkm_percent_cutoff/100)
  x_cut_i = which (x[,"rank"] < cutoff)
  x_cut = x[x_cut_i,]
  write.table(x_cut, file=outfile_top, quote = FALSE, sep="\t", row.names = FALSE)

  #Write out the subset of top N% genes that are in the GOI list
  goi_i=which(x_cut[,"mapped_gene_name"] %in% goi[[1]])
  if (length(goi_i) > 0){
    write.table(x_cut[goi_i,], file=outfile_top_goi, quote = FALSE, sep="\t", row.names = FALSE)
  }
}

writeFiles(genes_fpkm, genes_fpkm_expsort_file, genes_fpkm_expsort_goi_file, genes_fpkm_expsort_top_file, genes_fpkm_expsort_top_goi_file)
writeFiles(isoforms_fpkm, isoforms_fpkm_expsort_file, isoforms_fpkm_expsort_goi_file, isoforms_fpkm_expsort_top_file, isoforms_fpkm_expsort_top_goi_file)
writeFiles(isoforms_merged_fpkm, isoforms_merged_fpkm_expsort_file, isoforms_merged_fpkm_expsort_goi_file, isoforms_merged_fpkm_expsort_top_file, isoforms_merged_fpkm_expsort_top_goi_file)

#Create a distribution plot for the entire dataset


#Add ranks and percentiles to the dataset to be plotted
o = order(isoforms_merged_fpkm[,"FPKM"], na.last = TRUE, decreasing = TRUE)
plotdata = isoforms_merged_fpkm[o,]
plotdata[,"rank"]=1:length(plotdata[,1])
plotdata[,"percentile"]=round(((plotdata[,"rank"]/length(plotdata[,1]))*100), digits=3)

#Get the subset of data that is nonzero
fpkms=plotdata[,"FPKM"]
fpkms2= fpkms[which(fpkms>0)]


#Create plots for each of the genes of interest
setwd(images_subdir)
min_nonzero = 0.00001
for (gene_name in goi[[1]]){
  #print(gene_name)  
  image_file = paste(gene_name, "_hist.jpg", sep="")
  gene_i = which(plotdata[,"mapped_gene_name"] == gene_name)
  if (length(gene_i) == 1){
    x_label_text1 = paste("Gene expression (log2(FPKM + ", min_nonzero ,"))", sep="")
    x_label_text2 = "Gene expression (FPKM)"

    #jpeg(image_file, quality=100, width=800, height=1200, pointsize=15)
    #par(mfrow=c(2,1))
    
    jpeg(image_file, quality=100, width=800, height=800, pointsize=15)

    #Create a histogram of nonzero data and highlight the position of FLT3 - log2 scale
    hist(log2(fpkms2 + min_nonzero), breaks=100, col="light blue", xlab=x_label_text1, main="Distribution of non-zero gene expression values", font=2, font.lab=2)
    abline(v=log2(plotdata[gene_i,"FPKM"] + min_nonzero), col="red", lwd=2)
    percentile = plotdata[gene_i,"percentile"]
    legend_text = paste(gene_name, " (percentile = top ", percentile, "%)", sep="")
    legend("topleft", legend_text, col="red", lwd=2)

    #Create a histogram of nonzero data and highlight the position of FLT3 - normal scale
    #hist(fpkms2, breaks=100, col="light blue", xlab=x_label_text2, main="Distribution of non-zero gene expression values", font=2, font.lab=2)
    #abline(v=isoforms_merged_fpkm[gene_i,"FPKM"], col="red", lwd=2)
    #legend("topleft", gene_name, col="red", lwd=2)

    dev.off()
  }
}


